### PR TITLE
Add autofixer to `quotes` rule

### PR DESCRIPTION
**Pull request by courajs**
Waiting on:
- [x] [Merging quote-changing support to `ember-template-recast`](https://github.com/ember-template-lint/ember-template-recast/pull/653)
- [x] Release `ember-template-recast` including ^
- [x] Upgrade `ember-template-recast` version to `6.1.0`
- [x] Fixing the acceptance test which asserts on remaining unfixable rule violations. Of course I can just change this to have a different rule violation which doesn't currently have a fixer. But it seems better to have some always-failing rule for test purposes. Does anyone have ideas on the best way to add that? Preferably we wouldn't have to add an actual rule to `lib/rules`

Note: this won't autofix when escaping is required, blocked on: https://github.com/ember-template-lint/ember-template-recast/issues/698.

### CODE REVIEW

**Comment by dcyriller** [comment]
@courajs it is now possible to target `next` to leverage ember-template-recast@6.1.1.

**Comment by bmish** [comment]
@dcyriller I assume you mean ember-template-recast@6.1.0.

**Comment by dcyriller** [comment]
6.1.0 also contains a bugfix, hence the patch version in my mind. But you're totally right. I have updated my comment.

**Comment by courajs** [comment]
Ugh, fixing these acceptance tests is turning into a pain. But I'm determined! I will return to this later this week (•̀ᴗ•́)و

**Comment by courajs** [comment]
I declare this done!

File: test/acceptance/public-api-test.js [hunk 1]
```diff
@@ -286,31 +286,54 @@ describe('public api', function () {
...
         source: templateContents,
         filePath: templatePath,
         moduleId: templatePath.slice(0, -4),
       });
 
-      expect(result.messages).toEqual(expected);
-      expect(result.output).toEqual(templateContents);
-      expect(result.isFixed).toEqual(false);
+      expect(result).toMatchInlineSnapshot(
+        { messages: [{ filePath: expect.any(String) }] },
```
Comment (bmish) [inline test/acceptance/public-api-test.js:316]: This line looks wrong. I believe `toMatchInlineSnapshot` should only have one argument.

File: test/acceptance/public-api-test.js [hunk 2]
```diff
@@ -1394,31 +1417,54 @@ describe('public api', function () {
...
```
Comment (bmish) [inline test/acceptance/public-api-test.js:1447]: This line looks wrong. I believe `toMatchInlineSnapshot` should only have one argument.

File: test/unit/rules/quotes-test.js [hunk 3]
```diff
@@ -146,6 +169,7 @@ generateRuleTests({
     {
       config: 'single',
       template: '<input type="checkbox">',
```
Comment (bmish) [inline test/unit/rules/quotes-test.js:171]: Can we add test cases which combinations of different quote types?

```
<img alt="john's house">
```

I believe we would need to escape the quotes inside for this to fix properly:

```
<img alt='john\'s house'>
```

File: test/unit/rules/quotes-test.js [hunk 4]
```diff
@@ -33,9 +33,23 @@ generateRuleTests({
   ],
 
   bad: [
+    {
+      config: 'double',
+      template: "{{component 'one {{thing}} two'}}",
+      fixedTemplate: '{{component "one {{thing}} two"}}',
+
+      result: {
```
Comment (bmish) [inline test/unit/rules/quotes-test.js:41]: Can we use `verifyResults` instead so we can have a convenient snapshot?

**Review by bmish — CHANGES_REQUESTED** [review]

File: test/acceptance/public-api-test.js (hunk 2 above)
Comment (courajs) [reply test/acceptance/public-api-test.js:1447]: This is a jest feature called "property matchers" - it compensates for things in the snapshot that can change without causing a failure. In this case, the temporary file location is different every run, so this tells jest to just make sure they are both strings.
https://jestjs.io/docs/snapshot-testing#property-matchers

Comment (bmish) [reply test/acceptance/public-api-test.js:1447]: Oh, neat!

File: test/acceptance/public-api-test.js (hunk 1 above)
Comment (bmish) [reply test/acceptance/public-api-test.js:316]: Nevermind, see https://github.com/ember-template-lint/ember-template-lint/pull/2143/files#r758654423

File: test/unit/rules/quotes-test.js (hunk 3 above)
Comment (courajs) [reply test/unit/rules/quotes-test.js:171]: Oof, this is an excellent point. We'll actually need escape via html entities: https://html.spec.whatwg.org/dev/syntax.html#attributes-2

Comment (courajs) [reply test/unit/rules/quotes-test.js:171]: I think this [would be better fixed in `ember-template-recast`](https://github.com/ember-template-lint/ember-template-recast/issues/698)

File: test/unit/rules/quotes-test.js (hunk 4 above)
Comment (courajs) [reply test/unit/rules/quotes-test.js]: done

**Comment by courajs** [comment]
@bmish you raised an excellent issue about escapes, which I'm not really motivated to fix right now 😅 
Thanks for the review, but I think I'm gonna step away from this issue for now. Feel free to merge as-is, or close this PR.
If anyone else wants to get this over the finish line, you'd need to fix [this issue in `ember-template-recast`](https://github.com/ember-template-lint/ember-template-recast/issues/698)

**Comment by bmish** [comment]
@courajs how about you just disable the autofix for this special case now? We can still fix in the simple case where there's no need to do escaping. Then that more advanced case can be handled as a lower-priority follow-up.

**Review by bmish — APPROVED** [review]
Nice work!

File: lib/rules/quotes.js [hunk 5]
```diff
@@ -36,40 +36,76 @@ module.exports = class Quotes extends Rule {
...
 
     return {
       AttrNode(node) {
         if (!node.isValueless && node.quoteType === badChar) {
-          return this.log({
-            message,
-            node,
-          });
+          if (attrValueHasChar(node.value, goodChar)) {
+            return this.log({
```
Comment (bmish) [inline lib/rules/quotes.js:58]:
```suggestion
            // Autofix blocked on: https://github.com/ember-template-lint/ember-template-recast/issues/698
            return this.log({
```

File: lib/rules/quotes.js [hunk 6]
```diff
@@ -36,40 +36,76 @@ module.exports = class Quotes extends Rule {
...
 
-        if (source[0] === badChar) {
-          return this.log({
-            message,
-            node,
-            source: errorSource,
-          });
+        if (node.quoteType === badChar) {
+          if (node.value.includes(goodChar)) {
+            return this.log({
```
Comment (bmish) [inline lib/rules/quotes.js:80]:
```suggestion
            // Autofix blocked on: https://github.com/ember-template-lint/ember-template-recast/issues/698
            return this.log({
```

File: test/unit/rules/quotes-test.js [hunk 7]
```diff
@@ -165,6 +201,72 @@ generateRuleTests({
         `);
       },
     },
+    {
+      config: 'single',
```
Comment (bmish) [inline test/unit/rules/quotes-test.js:205]:
```suggestion
      // Autofix blocked on: https://github.com/ember-template-lint/ember-template-recast/issues/698
      config: 'single',
```

File: test/unit/rules/quotes-test.js [hunk 8]
```diff
@@ -165,6 +201,72 @@ generateRuleTests({
...
+              "rule": "quotes",
+              "severity": 2,
+              "source": "class='a \\"so-called\\" btn {{this.otherClass}}'",
+            },
+          ]
+        `);
+      },
+    },
+    {
+      config: 'single',
```
Comment (bmish) [inline test/unit/rules/quotes-test.js:249]:
```suggestion
      // Autofix blocked on: https://github.com/ember-template-lint/ember-template-recast/issues/698
      config: 'single',
```

File: test/unit/rules/quotes-test.js [hunk 9]
```diff
@@ -165,6 +201,72 @@ generateRuleTests({
...
+              "rule": "quotes",
+              "severity": 2,
+              "source": "alt=\\"Abdul's house\\"",
+            },
+          ]
+        `);
+      },
+    },
+    {
+      config: 'double',
```
Comment (bmish) [inline test/unit/rules/quotes-test.js:227]:
```suggestion
      // Autofix blocked on: https://github.com/ember-template-lint/ember-template-recast/issues/698
      config: 'double',
```

**Review by bmish — APPROVED** [review]

**Pull request merged**